javascript/test/tokenizeTest.ts

```diff
@@ -1,121 +1,137 @@
 import assert from 'assert'
-import { isSetIterator } from 'util/types'
 
 import { Token, TokenType } from '../src/Ast.js'
 
-const tokenize: (input: string) => Token[] = (input) => {
-  const tokens: Array<Token> = []
+class Cursor {
+  constructor(public readonly input: string, public readonly currentIndex: number) {}
 
-  //  "hello world" --> 3 tokens
-  //  "hello  world"  --> 4 tokens
-  //       ^
-  // firstIndex
-  // curentType
-
-  let startOfWord = -1
-  let currentIndex = 0
-  let tokenType: TokenType | undefined = undefined
-  let previousTokenType: TokenType | undefined
-
-  for (currentIndex; currentIndex < input.length; currentIndex++) {
-    previousTokenType = tokenType
-    if (input[currentIndex] === ' ') {
-      tokenType = TokenType.whiteSpace
-    } else {
-      tokenType = TokenType.text
+  get tokenType(): TokenType {
+    return Token.typeOf(this.input[this.currentIndex])
+  }
+
+  get previousTokenType(): TokenType | undefined {
+    if (this.currentIndex === 0) {
+      return undefined
     }
+    return Token.typeOf(this.input[this.currentIndex - 1])
+  }
+
+  get atStartOfWord(): boolean {
+    return this.previousTokenType !== TokenType.text && this.tokenType === TokenType.text
+  }
+
+  get atEndOfWord(): boolean {
+    return this.previousTokenType === TokenType.text && this.tokenType !== TokenType.text
+  }
 
-    // moving into a string?
-    if (previousTokenType !== TokenType.text && tokenType === TokenType.text) {
-      startOfWord = currentIndex
-    } 
-    
-    if (
-      previousTokenType !== undefined &&
-      previousTokenType !== tokenType &&
-      tokenType !== TokenType.text
-    ) {
-      const subString = input.slice(startOfWord, currentIndex)
-      tokens.push(new Token(TokenType.text, subString, startOfWord, startOfWord + subString.length))
+  get atEndOfInput(): boolean {
+    return this.currentIndex == this.input.length
+  }
+
+  get endOfCurrentWord(): number {
+    let cursor = new Cursor(this.input, this.currentIndex)
+    while (!cursor.atEndOfInput) {
+      cursor = new Cursor(this.input, cursor.currentIndex + 1)
+      if (cursor.atEndOfWord) {
+        return cursor.currentIndex
+      }
     }
+    return this.input.length
+  }
 
-    if (tokenType !== TokenType.text) {
-      tokens.push(new Token(tokenType, input[currentIndex], currentIndex, currentIndex + 1))
+  scan(emit: (token: Token) => void): Cursor {
+    if (this.atStartOfWord) {
+      const word = this.input.slice(this.currentIndex, this.endOfCurrentWord)
+      emit(new Token(TokenType.text, word, this.currentIndex, this.endOfCurrentWord))
+      return new Cursor(this.input, this.endOfCurrentWord)
     }
+
+    emit(
```

Comment:
Not every language supports generators. We have to implement this in Go and Java too. Which will be easier if we use a smaller subset of language features.

Additionally this creates quite a bit of pressure on the garbage collector by creating disposable objects in a loop.